[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module

yhcharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90711

To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

yhcharles committed:
[Composable API] Check for `fully_shard` in `replicate` API
To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

rohan-varma commented on test/distributed/_composable/test_compose.py:
```diff
@@ -171,6 +179,37 @@ def test_checkpoint_fsdp_submodules_with_param_no_shard(self):
+
+    @skip_if_lt_x_gpu(2)
+    def test_replicate_in_fsdp(self):
+        # TODO: to be added
```
In general I feel like it's better to omit this and file an issue, or write out what the test might be like and use `assertRaises` for the parts that fail.

rohan-varma commented on test/distributed/_composable/test_compose.py:
```diff
@@ -171,6 +179,37 @@ def test_checkpoint_fsdp_submodules_with_param_no_shard(self):
+            AssertionError, "Cannot apply .*replicate"
+        ):
+            replicate(model.l1)
+        replicate(model.l2)  # should not raise
```
would it be useful to run a forward and backward pass to validate?

yhcharles replied:
Thanks!
This test only verifies how the APIs can be composed. For example, if apply both `fully_shard` and `replicate` on the same module, it should raise exception. Let me add a comment here.
The correctness of `forward` and `backward` are verified in the following test.

rohan-varma commented on test/distributed/_composable/test_compose.py:
```diff
@@ -171,6 +179,37 @@ def test_checkpoint_fsdp_submodules_with_param_no_shard(self):
+        model = CompositeModel(device=torch.device("cuda"))
+        base_model = copy.deepcopy(model)
+        test_model = copy.deepcopy(model)
+        fully_shard(test_model.l1)
```
Should we tested nested/wrapped FSDP as well?

yhcharles replied:
@rohan-varma do you mean test `replicate` with FSDP wrapper?

rohan-varma replied:
SOrry should've clarified, meant passing a policy like @awgu suggests

mrshenli replied:
+1 on adding the test, but if there are gaps there, we don't have to fix it in this PR. Having a test with `assertRaisesRegex` will be sufficient for this one, and we can fix in follow up PRs.

rohan-varma reviewed (commented):
LGTM, just a few comments on testing.

awgu commented on test/distributed/_composable/test_compose.py:
```diff
@@ -171,6 +179,37 @@ def test_checkpoint_fsdp_submodules_with_param_no_shard(self):
             self._test_parity,
         )
 
+    def test_composable_fsdp_replicate(self):
```
```suggestion
    @skip_if_lt_x_gpu(2)
    def test_composable_fsdp_replicate(self):
```

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/_composable/replicate.py:
```diff
@@ -25,8 +51,11 @@ def mark_modules(self, *modules: nn.Module, **kwargs) -> None:
     def _recursive_collect_params(self, module: nn.Module) -> None:
-        # TODO: skip if managed by other APIs
+        # skip if managed by other APIs
+        if not _composable(module):
```
naming wise, should this be specific to FSDP or other parallelism APIs? If we have activation checkpoint, that should be fine here, right?

yhcharles replied:
You're right. `fully_shard` and `replicate` cannot be used on the same module, but activation checkpointing and replicate can.

awgu replied:
In that case, I would be okay being explicit and calling the function `_is_composable_with_fully_shard()`. I plan to call the corresponding function `_is_composable_with_replicate()`.

yhcharles replied:
I will prefer to keep the name as it is. The reason is we will probably have similar function in each of `replicate.py` and `fully_shard.py`, so the full path of the function name will be `replicate._composable()` or `fully_shard._composable()`. And in each of the implementation, the `_composable()` function will check if a module can be marked with the API in current file. Does this make sense? @awgu @rohan-varma

awgu replied:
That sounds good to me.

awgu replied:
One minor thing to mention: In FSDP code right now, we have a helper `_is_composable(state: _FSDPState)` that is used for now to differentiate between the wrapper and composable code paths.

To avoid confusion with `_composable`, should we rename `_composable()` to `_can_compose()`? It is not a big deal either way; in the FSDP code, we just need to be careful everyone knows what `_is_composable` means vs. `_composable`.

cc: @rohan-varma

awgu replied:
Actually, I kind of like `_composable` as well. Up to you guys.

yhcharles replied:
Updated to `_can_compose`

rohan-varma commented on test/distributed/_composable/test_compose.py:
```diff
@@ -76,7 +76,9 @@ def test_wrap_same_submodule(self, use_reentrant: bool):
         # compose checkpoint and fully_shard
-        test_model.seq = checkpoint(test_model.seq, use_reentrant=use_reentrant)
+        test_model.seq = checkpoint(
+            test_model.seq, use_reentrant=use_reentrant
```
nit: let's try to avoid formatting changes and logical changes in the same PR as it makes it harder to review.

rohan-varma commented on test/distributed/_composable/test_compose.py:
```diff
@@ -159,8 +165,39 @@ def test_checkpoint_fsdp_submodules_with_param_no_shard(self):
+        model = CompositeModel(device=torch.device("cuda"))
+        base_model = copy.deepcopy(model)
+        test_model = copy.deepcopy(model)
+        fully_shard(test_model.l1)
```
Same comment about testing with auto_wrap.

mrshenli commented on torch/distributed/_composable/replicate.py:
```diff
@@ -4,7 +4,30 @@
+    module: nn.Module,  # NOTE: contract now supports single module only
+    **kwargs,
+) -> nn.Module:
+    r"""Replicates a module
+
+    Args:
+        module (torch.nn.Module): module to replicate
+
+    Example::
+        >>> module = nn.Linear(3, 3)
+        >>> replicate(module)
+    """
```
can be done in followup PRs, but we will need a much more comprehensive API docstring, e.g., what does this do, what are expected kwargs, how does this compose with others, and what are mutual-exclusive APIs with this one, etc.

mrshenli reviewed (commented):
Hey @yhcharles, are there any blocker for landing this PR?

mrshenli reviewed (approved):
Stamp to unblock, please address comments from @rohan-varma and @awgu

awgu commented on torch/distributed/_composable/replicate.py:
```diff
@@ -25,8 +51,11 @@ def mark_modules(self, *modules: nn.Module, **kwargs) -> None:
+        if not _composable(module):
+            return
 
+        # skip if module parameters already collected
         if hasattr(replicate.state(module), "_params_collected"):
             if replicate.state(module)._params_collected:
```
nit: Before this PR:
```suggestion
        # skip if module parameters already collected
        if getattr(replicate.state(module), "_params_collected", False):
```

yhcharles replied:
Thanks, it's simpler :) But with L62, it seems we can't do this?

awgu replied:
Oh sorry, I did not see below. Definitely keep it as is :)

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/replicate.py:
```diff
@@ -16,6 +39,9 @@ def __init__(self) -> None:
 
     def mark_modules(self, *modules: nn.Module, **kwargs) -> None:
         for module in modules:
+            assert _can_compose(
```
nit: I am wondering if we should make this an explicit `raise AssertionError` since `assert` may be optimized out.

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

yhcharles commented:
@pytorchbot merge -g

yhcharles committed:
Update on "[Composable API] Enable composable `fully_shard` submodules in `replicate` parent module"


To make sure `fully_shard` and `replicate` can work together, we need to check for each other in the implementation. This change adds the check in `replicate()`.

[ghstack-poisoned]

yhcharles commented:
@pytorchbot merge -g
